ENH: use studentized range over statsmodels for pairwise tests

swallan opened the pull request:
This PR replaces use of statsmodels' studentized range distribution functions (`statsmodels.stats.libqsturng.qsturng`, `statsmodels.stats.libqsturng.psturng`) with those from SciPy now that `scipy.stats.studentized_range` is available (SciPy 1.7.0+). 

Speed is maintained, and results are more accurate because SciPy's implementation evaluates the integral for the provided arguments rather than interpolating from tabulated results. statsmodels [has adopted](https://github.com/statsmodels/statsmodels/pull/8035) this change internally as well.

For example, this box and whiskey plot (from [here](https://zenodo.org/record/5151976#.YekY3_7MKUk)) compares the relative accuracy of the studentized range CDF in SciPy, statsmodels, and R (ptukey) across a wide range of inputs.


<img width="868" alt="Screen Shot 2022-01-24 at 9 27 21 PM" src="https://user-images.githubusercontent.com/44255917/150916550-99966463-89d5-4419-9f08-e247bc5c148f.png">

raphaelvallat commented:
This is fantastic, thanks @swallan! I somehow missed that addition in SciPy. The unit tests are failing because of an unrelated issue (https://github.com/raphaelvallat/pingouin/issues/227), but have you tried to run the test and doctest locally? Are there any modifications required? I am assuming that the overall impact on the resulting p-values is going to be so small that it might not even require updating the unit tests and examples in the docstring, but could you please check?

Also, could you edit the [changelog.rst](https://github.com/raphaelvallat/pingouin/blob/master/docs/changelog.rst) file, under the 0.6.0.dev section?

PS: read more about the contributing guidelines [here](https://pingouin-stats.org/contributing.html).

Thank you!

swallan commented:
@raphaelvallat I did run the `pytest` suite locally after switching pandas versions and it was all successful.

Output of locally running tests:
<details>
<pre>
❯ pytest pingouin
================================================= test session starts =================================================
platform darwin -- Python 3.8.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/swallan/Documents/SciPy/scipy_meson/pingouin, configfile: setup.cfg
collected 86 items                                                                                                    

pingouin/tests/test_bayesian.py ...                                                                             [  3%]
pingouin/tests/test_circular.py ........                                                                        [ 12%]
pingouin/tests/test_config.py .                                                                                 [ 13%]
pingouin/tests/test_contingency.py ....                                                                         [ 18%]
pingouin/tests/test_correlation.py ....                                                                         [ 23%]
pingouin/tests/test_distribution.py ......                                                                      [ 30%]
pingouin/tests/test_effsize.py .....                                                                            [ 36%]
pingouin/tests/test_equivalence.py .                                                                            [ 37%]
pingouin/tests/test_multicomp.py .....                                                                          [ 43%]
pingouin/tests/test_multivariate.py ...                                                                         [ 46%]
pingouin/tests/test_nonparametric.py ........                                                                   [ 55%]
pingouin/tests/test_pairwise.py ....                                                                            [ 60%]
pingouin/tests/test_pandas.py .                                                                                 [ 61%]
pingouin/tests/test_parametric.py .......                                                                       [ 69%]
pingouin/tests/test_plotting.py .......                                                                         [ 77%]
pingouin/tests/test_power.py ......                                                                             [ 84%]
pingouin/tests/test_regression.py ...                                                                           [ 88%]
pingouin/tests/test_reliability.py ..                                                                           [ 90%]
pingouin/tests/test_utils.py ........                                                                           [100%]

================================================== warnings summary ===================================================
venv/lib/python3.8/site-packages/seaborn/rcmod.py:82
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/seaborn/rcmod.py:82: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if LooseVersion(mpl.__version__) >= "3.0":

venv/lib/python3.8/site-packages/setuptools/_distutils/version.py:351
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/setuptools/_distutils/version.py:351: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    other = LooseVersion(other)

venv/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2631
pingouin/tests/test_parametric.py::TestParametric::test_anova
pingouin/tests/test_parametric.py::TestParametric::test_rm_anova2
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2631: FutureWarning: The `inplace` parameter in pandas.Categorical.add_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.
    res = method(*args, **kwargs)

venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:35
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:35: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    version = LooseVersion(pd.__version__)

venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:37
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    pandas_lt_1_0_0 = version < LooseVersion("1.0.0")

venv/lib/python3.8/site-packages/statsmodels/compat/numpy.py:46
venv/lib/python3.8/site-packages/statsmodels/compat/numpy.py:46
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/numpy.py:46: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    NP_LT_114 = LooseVersion(np.__version__) < LooseVersion("1.14")

venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:6
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:6: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SP_VERSION = LooseVersion(scipy.__version__)

venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:7
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:7: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SCIPY_GT_14 = SP_VERSION >= LooseVersion("1.5")

venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:8
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:8: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SP_LT_15 = SP_VERSION < LooseVersion("1.5")

venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:9
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:9: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SP_LT_16 = SP_VERSION < LooseVersion("1.6")

pingouin/tests/test_correlation.py::TestCorrelation::test_corr
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
    warnings.warn(PearsonRConstantInputWarning())

pingouin/tests/test_equivalence.py::TestEquivalence::test_tost
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/pingouin/bayesian.py:146: RuntimeWarning: invalid value encountered in double_scalars
    bf10 = 1 / ((1 + t**2 / df)**(-(df + 1) / 2) / integr)

pingouin/tests/test_equivalence.py::TestEquivalence::test_tost
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/pingouin/bayesian.py:146: RuntimeWarning: divide by zero encountered in double_scalars
    bf10 = 1 / ((1 + t**2 / df)**(-(df + 1) / 2) / integr)

pingouin/tests/test_equivalence.py::TestEquivalence::test_tost
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/pingouin/parametric.py:263: RuntimeWarning: invalid value encountered in multiply
    ci = np.array([tval - tcrit, tval + tcrit]) * se

-- Docs: https://docs.pytest.org/en/stable/warnings.html
================================================ slowest 10 durations =================================================
9.29s call     pingouin/tests/test_effsize.py::TestEffsize::test_compute_boot_esci
6.51s call     pingouin/tests/test_regression.py::TestRegression::test_mediation_analysis
1.48s call     pingouin/tests/test_pairwise.py::TestPairwise::test_pairwise_corr
1.43s call     pingouin/tests/test_pairwise.py::TestPairwise::test_pairwise_ttests
1.25s call     pingouin/tests/test_plotting.py::TestPlotting::test_plot_paired
0.57s call     pingouin/tests/test_pandas.py::TestParametric::test_pandas
0.50s call     pingouin/tests/test_plotting.py::TestPlotting::test_plot_rm_corr
0.42s call     pingouin/tests/test_parametric.py::TestParametric::test_mixed_anova
0.33s call     pingouin/tests/test_plotting.py::TestPlotting::test_plot_blandaltman
0.30s call     pingouin/tests/test_plotting.py::TestPlotting::test_plot_shift
========================================== 86 passed, 17 warnings in 30.41s ===========================================
</pre>
</details>

> impact on the resulting p-values is going to be so small

I didn't realize running the doctests before, but I have now updated the p-values. The changes needed were small. Here are the tests passing with that change (to be pushed): 

<details>
<pre>
❯ pytest --doctest-modules
============================================================== test session starts ==============================================================
platform darwin -- Python 3.8.9, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /Users/swallan/Documents/SciPy/scipy_meson/pingouin, configfile: setup.cfg
collected 165 items                                                                                                                             

bayesian.py ...                                                                                                                           [  1%]
circular.py ........                                                                                                                      [  6%]
contingency.py ...                                                                                                                        [  8%]
correlation.py ......                                                                                                                     [ 12%]
distribution.py ......                                                                                                                    [ 15%]
effsize.py .....                                                                                                                          [ 18%]
equivalence.py .                                                                                                                          [ 19%]
multicomp.py .....                                                                                                                        [ 22%]
multivariate.py ...                                                                                                                       [ 24%]
nonparametric.py ........                                                                                                                 [ 29%]
pairwise.py ....                                                                                                                          [ 31%]
parametric.py ......                                                                                                                      [ 35%]
plotting.py ......                                                                                                                        [ 38%]
power.py ......                                                                                                                           [ 42%]
regression.py ...                                                                                                                         [ 44%]
reliability.py ..                                                                                                                         [ 45%]
utils.py ..                                                                                                                               [ 46%]
datasets/__init__.py ..                                                                                                                   [ 47%]
tests/test_bayesian.py ...                                                                                                                [ 49%]
tests/test_circular.py ........                                                                                                           [ 54%]
tests/test_config.py .                                                                                                                    [ 55%]
tests/test_contingency.py ....                                                                                                            [ 57%]
tests/test_correlation.py ....                                                                                                            [ 60%]
tests/test_distribution.py ......                                                                                                         [ 63%]
tests/test_effsize.py .....                                                                                                               [ 66%]
tests/test_equivalence.py .                                                                                                               [ 67%]
tests/test_multicomp.py .....                                                                                                             [ 70%]
tests/test_multivariate.py ...                                                                                                            [ 72%]
tests/test_nonparametric.py ........                                                                                                      [ 76%]
tests/test_pairwise.py ....                                                                                                               [ 79%]
tests/test_pandas.py .                                                                                                                    [ 80%]
tests/test_parametric.py .......                                                                                                          [ 84%]
tests/test_plotting.py .......                                                                                                            [ 88%]
tests/test_power.py ......                                                                                                                [ 92%]
tests/test_regression.py ...                                                                                                              [ 93%]
tests/test_reliability.py ..                                                                                                              [ 95%]
tests/test_utils.py ........                                                                                                              [100%]

=============================================================== warnings summary ================================================================
../venv/lib/python3.8/site-packages/seaborn/rcmod.py:82
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/seaborn/rcmod.py:82: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if LooseVersion(mpl.__version__) >= "3.0":

../venv/lib/python3.8/site-packages/setuptools/_distutils/version.py:351
pingouin/plotting.py::pingouin.plotting.plot_circmean
pingouin/plotting.py::pingouin.plotting.plot_rm_corr
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/setuptools/_distutils/version.py:351: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    other = LooseVersion(other)

../venv/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2631
pingouin/tests/test_parametric.py::TestParametric::test_anova
pingouin/tests/test_parametric.py::TestParametric::test_rm_anova2
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/pandas/core/arrays/categorical.py:2631: FutureWarning: The `inplace` parameter in pandas.Categorical.add_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.
    res = method(*args, **kwargs)

../venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:35
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:35: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    version = LooseVersion(pd.__version__)

../venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:37
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/pandas.py:37: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    pandas_lt_1_0_0 = version < LooseVersion("1.0.0")

../venv/lib/python3.8/site-packages/statsmodels/compat/numpy.py:46
../venv/lib/python3.8/site-packages/statsmodels/compat/numpy.py:46
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/numpy.py:46: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    NP_LT_114 = LooseVersion(np.__version__) < LooseVersion("1.14")

../venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:6
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:6: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SP_VERSION = LooseVersion(scipy.__version__)

../venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:7
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:7: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SCIPY_GT_14 = SP_VERSION >= LooseVersion("1.5")

../venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:8
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:8: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SP_LT_15 = SP_VERSION < LooseVersion("1.5")

../venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:9
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/statsmodels/compat/scipy.py:9: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    SP_LT_16 = SP_VERSION < LooseVersion("1.6")

pingouin/plotting.py::pingouin.plotting.plot_circmean
pingouin/plotting.py::pingouin.plotting.plot_rm_corr
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/seaborn/rcmod.py:400: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if LooseVersion(mpl.__version__) >= "3.0":

pingouin/plotting.py::pingouin.plotting.plot_rm_corr
pingouin/plotting.py::pingouin.plotting.plot_rm_corr
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/seaborn/axisgrid.py:130: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if LooseVersion(mpl.__version__) < LooseVersion("3.0"):

pingouin/tests/test_correlation.py::TestCorrelation::test_corr
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/venv/lib/python3.8/site-packages/scipy/stats/stats.py:4023: PearsonRConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
    warnings.warn(PearsonRConstantInputWarning())

pingouin/tests/test_equivalence.py::TestEquivalence::test_tost
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/pingouin/bayesian.py:146: RuntimeWarning: invalid value encountered in double_scalars
    bf10 = 1 / ((1 + t**2 / df)**(-(df + 1) / 2) / integr)

pingouin/tests/test_equivalence.py::TestEquivalence::test_tost
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/pingouin/bayesian.py:146: RuntimeWarning: divide by zero encountered in double_scalars
    bf10 = 1 / ((1 + t**2 / df)**(-(df + 1) / 2) / integr)

pingouin/tests/test_equivalence.py::TestEquivalence::test_tost
  /Users/swallan/Documents/SciPy/scipy_meson/pingouin/pingouin/parametric.py:263: RuntimeWarning: invalid value encountered in multiply
    ci = np.array([tval - tcrit, tval + tcrit]) * se

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================= slowest 10 durations ==============================================================
10.04s call     pingouin/tests/test_effsize.py::TestEffsize::test_compute_boot_esci
6.90s call     pingouin/tests/test_regression.py::TestRegression::test_mediation_analysis
2.76s call     pingouin/tests/test_plotting.py::TestPlotting::test_plot_paired
2.06s call     pingouin/regression.py::pingouin.regression.mediation_analysis
1.67s call     pingouin/tests/test_pairwise.py::TestPairwise::test_pairwise_ttests
1.64s call     pingouin/tests/test_pairwise.py::TestPairwise::test_pairwise_corr
1.58s call     pingouin/effsize.py::pingouin.effsize.compute_bootci
0.96s call     pingouin/tests/test_plotting.py::TestPlotting::test_plot_rm_corr
0.69s call     pingouin/plotting.py::pingouin.plotting.plot_rm_corr
0.57s call     pingouin/tests/test_pandas.py::TestParametric::test_pandas
======================================================= 165 passed, 23 warnings in 39.72s =======================================================
</pre>
</details>

I also updated the changelog.

raphaelvallat commented on pingouin/pairwise.py:
```diff
@@ -630,11 +631,10 @@ def pairwise_tukey(data=None, dv=None, between=None, effsize='hedges'):
-    1     Adelie     Gentoo  3700.662  5076.016 -1375.354  56.148 -24.495    0.001  -2.967
-    2  Chinstrap     Gentoo  3733.088  5076.016 -1342.928  69.857 -19.224    0.001  -2.894
+    0     Adelie  Chinstrap  3700.662  3733.088   -32.426  67.512  -0.480    0.881  -0.070
+    1     Adelie     Gentoo  3700.662  5076.016 -1375.354  56.148 -24.495   -0.000  -2.967
```
@swallan why the negative sign here? a p-value cannot be negative.

swallan replied:
Absolutely right. I took a look at the parameters that are being used to get the p-values and evaluated them with mpmath, and it looks like the true result of what they should be is `mpf('0.0')`. (I spun up [this notebook](https://colab.research.google.com/drive/1PlgjaDbBKGl67QPOY5BDO_VDRp4bCpJ4?usp=sharing) to show this)

The result of evaluation with `scipy.stats.studentized_range` for those suspicious negative p-values are `-9.9253938401489e-14`. What we can do here is `np.clip(pvalues, 0, 1)`, to ensure they fall within the range while I submit a PR to scipy to include this behavior by default.

raphaelvallat replied:
This is great, thanks!

raphaelvallat replied:
Could you please update the docstring and then I think we'll be good to merge?

swallan replied:
Done! When [this pr](https://github.com/scipy/scipy/pull/15525) makes it into a SciPy release and pingouin's minimum SciPy version includes that feature I'll follow up to remove the unneeded clipping.

swallan replied:
@raphaelvallat Just wanted to check in -- is there anything else that you would like taken care of in this PR or is it merge ready?

raphaelvallat commented:
Merging now, thanks again!
